analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/CreatingHashAlgorithmsBase.cs

```diff
@@ -28,23 +28,50 @@ public abstract class CreatingHashAlgorithmsBase<TSyntaxKind> : SonarDiagnosticA
+        {
+            KnownType.System_Security_Cryptography_SHA1,
+            KnownType.System_Security_Cryptography_MD5,
+            KnownType.System_Security_Cryptography_DSA,
+            KnownType.System_Security_Cryptography_RIPEMD160,
+            KnownType.System_Security_Cryptography_HMACSHA1,
+            KnownType.System_Security_Cryptography_HMACMD5,
+            KnownType.System_Security_Cryptography_HMACRIPEMD160
+        };
+
+        private readonly string[] unsafeAlgorithms = { "SHA1", "MD5", "DSA", "HMACMD5", "HMACRIPEMD160", "RIPEMD160", "RIPEMD160Managed" };
+
         protected ObjectCreationTracker<TSyntaxKind> ObjectCreationTracker { get; set; }
 
-        protected InvocationTracker<TSyntaxKind> InvocationTracker { get; set;  }
+        protected InvocationTracker<TSyntaxKind> ParameterlessFactoryInvocationTracker { get; set; }
 
-        protected BaseTypeTracker<TSyntaxKind> BaseTypeTracker { get; set; }
+        protected InvocationTracker<TSyntaxKind> ParameterizedFactoryInvocationTracker { get; set; }
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            ObjectCreationTracker.Track(context,
-                ObjectCreationTracker.WhenDerivesFrom(KnownType.System_Security_Cryptography_HashAlgorithm));
+            ObjectCreationTracker.Track(context, ObjectCreationTracker.WhenDerivesOrImplementsAny(algorithmTypes));
 
-            InvocationTracker.Track(context,
-                InvocationTracker.MethodNameIs("Create"),
-                InvocationTracker.MethodReturnTypeIs(KnownType.System_Security_Cryptography_HashAlgorithm));
+            ParameterlessFactoryInvocationTracker.Track(context,
+                                                        ParameterlessFactoryInvocationTracker.MatchMethod(new MemberDescriptor(KnownType.System_Security_Cryptography_DSA, Create),
+                                                                                                          new MemberDescriptor(KnownType.System_Security_Cryptography_HMAC, Create),
+                                                                                                          new MemberDescriptor(KnownType.System_Security_Cryptography_MD5, Create),
+                                                                                                          new MemberDescriptor(KnownType.System_Security_Cryptography_RIPEMD160, Create),
+                                                                                                          new MemberDescriptor(KnownType.System_Security_Cryptography_SHA1, Create)),
+                                                        ParameterlessFactoryInvocationTracker.MethodHasParameters(0));
 
-            BaseTypeTracker.Track(context,
-                BaseTypeTracker.MatchSubclassesOf(KnownType.System_Security_Cryptography_HashAlgorithm));
+            ParameterizedFactoryInvocationTracker.Track(context,
```

Comment:
Same as before, can we use lambda to avoid repetition? With one additional `"CreateFromName"` concat.

Reply:
I prefer not to do this change since not all of them have a create method.